### ISSUE TITLE
Figure out how to handle frame callbacks on Wayland

### ISSUE THREAD

**Issue by emersion** [glfw/glfw#1350]
On Wayland, when a window is hidden it won't receive frame events. This will block `eglSwapBuffers` if the window is hidden, which potentially blocks the whole application.

See https://github.com/swaywm/sway/issues/2770

**Comment by linkmauve** [comment]
I think the approach applications should usually use is to handle the winsys (here GLFW) on one thread, and do the OpenGL calls (including `glfwSwapBuffers()`) on another thread, avoiding all kind of issues where the rendering thread is blocking the rest of the application.  Most programs I know which aren’t toy projects are using this approach for performances.

A better story using both `wl_buffer.frame` and `wp_presentation` would need to be drafted, but I still don’t know how to expose it correctly to the user.

**Comment by ewtoombs** [comment]
I thought about this for a while. I think the best way to use `wp_presentation` without breaking the semantics of the rest of the API is to make a new function, which I'll provisionally call `glfwSwapPunctual()`. This is what it does when called:

1. Set the `wp_presentation` callback to return when the new front buffer is presented.
2. Swap the buffers using `eglSwapBuffers()`, so the most recently rendered buffer is now in the front.
3. Wait until the callback is received.
4. Return the time when the new front buffer is displayed. (In seconds as a `double` such that it agrees with `glfwGetTime()`.)

This function will only work properly if `eglSwapInterval` is set to 0.

This function should be implementable on any platform that can support the same semantics.

It is called punctual because at least in weston it should eliminate a whole frame of latency. This is because the current `eglSwapBuffers()` with `eglSwapInterval(1)` blocks until a few milliseconds before the *old* front buffer is displayed, *then* swaps the buffers and commits the new front buffer. I think this is right. Correct me if I'm wrong, @emersion . Also, hi, emersion! Fancy seeing you here!

I am going to implement this function as I have described it, just to make sure I'm right about this. When it's done, I'll send screenshots.

**Comment by emersion** [comment]
I don't think it's a good way to fix this issue. I'm not sure the presentation-time protocol is needed at all tbh (it's more useful for video players).

Frame callbacks should be used in the general case, that's for sure. Blocking is never a good idea, so it would be better to expose some kind of frame signal in the GLFW API. It would be interesting to see if other platforms have different ways to address this issue.

**Comment by ewtoombs** [comment]
Why would you insist on a non-blocking API when multiplexing so few streams?

**Comment by ewtoombs** [comment]
My own goal in this is to reduce the latency to the minimum possible amount, which is best case T_R, worst case T_F + T_R, where T_R is the amount of time it takes to render a frame and T_F is the amount of time in between frames. GLFW is not currently capable of this. However you decide to implement this is fine with me, really. It's worth noting, though, that the GLFW API is already blocking.

**Comment by emersion** [comment]
If the compositor gives a render budget for each client (could be dynamic), then gives itself a budget to composite, you'll get a frame event after the presentation-time event (ie. better latency) that'll be presented on the next vblank.

**Comment by ewtoombs** [comment]
The compositor has no idea how long I am going to take to render. I need to be able to determine my own budget.

**Comment by emersion** [comment]
That's why the budget can be dynamic.

If you want to manage your own budget, you _should_ use presentation-time. However this requires special support in clients. Thus it should be a totally optional API.

**Comment by ewtoombs** [comment]
That's why I made a new function call.

**Comment by ewtoombs** [comment]
Also, the glfw event loop will have to keep running while waiting for the presentation callback in order to receive it at the right time, so all of glfw's callbacks will keep working.

**Comment by ewtoombs** [comment]
The basic idea is
```
while(!windowShouldClose()) {
    render();
    swapPunctual();
    processEventsFor(t_frame - t_render);
}
```
where `processEventsFor(t)` processes GLFW events for `t` seconds, `t_frame` is the time in between frames, and `t_render` is my rendering budget.

**Comment by emersion** [comment]
New idea: `glfwSwapBuffers` could implement its own frame callback logic to block while still processing other events.

**Comment by ewtoombs** [comment]
Yeah, I guess that could happen. I mean glfwSwapBuffers() is already semantically different between X11 and wayland. X11 never blocks and wayland does. X11 blocks on the next GL rendering call instead. This could just be the new way we decide to swap buffers. It might even be possible to get the same semantics to work in X11.

**Comment by ewtoombs** [comment]
There is one more thing to consider, though. swapPunctual() would return the presentation time. Getting glfwSwapBuffers() to do that would break its ABI.

**Comment by ewtoombs** [comment]
There is also the interaction with glfwSwapInterval() to consider. To get consistent behaviour across platforms, we would have to handle the interval ourselves, while actually setting eglSwapInterval() to 0.

**Comment by ewtoombs** [comment]
I just got presentation-time working. The results are as predicted. I have used it to synchronise glfw perfectly with the compositor's mouse. I put my changes here: https://github.com/glfw/glfw/pull/1406
